File: packages/phonetics-core/src/atlas/load-atlas.ts
/**
 * Phonetic Atlas Loader — TypeScript entry point for W2
 * =========================================================
 *
 * Loads the compiled FlatBuffers Phonetic Atlas (.bin) into an
 * ArrayBuffer and provides zero-copy access to phoneme feature vectors.
 *
 * WHY a separate loader module:
 *   - The FlatBuffers binary is a static asset checked into the repo
 *     (see architecture.md §5). This loader encapsulates the binary
 *     load, buffer verification (file identifier), and provides a
 *     convenient lookup API with NFC-normalized index.
 *   - Zero-copy: FeatureVector is a 'struct' (fixed 24 bytes inline).
 *     Reading a single field does NOT parse the entire buffer — the
 *     data is already at the correct offset.
 *
 * IPA NORMALIZATION POLICY:
 *   ALL IPA strings are normalized to NFC (Normalization Form Canonical
 *   Composition) on both index build and lookup. This ensures that
 *   a query for "ä" (U+00E4, precomposed) finds the same entry as
 *   "a\u0308" (U+0061 + U+0308, decomposed). Both the generator
 *   (Python) and this loader (TypeScript) must agree on NFC.
 *
 * USAGE:
 *   import { PhoneticAtlasIndex } from './atlas/load-atlas';
 *   const atlas = PhoneticAtlasIndex.fromBuffer(response.arrayBuffer());
 *   const phoneme = atlas.get('p');
 *   if (phoneme) {
 *     console.log(phoneme.ipa, phoneme.features.voi());
 *   }
 *
 * CONSUMPTION PATTERN (per architecture.md §5):
 *   - W2 (TS)       → this module, loaded once at worker init
 *   - Python        → load_atlas.py (flatc Python bindings + mmap)
 *   - Rust FST (W1) → native FlatBuffers Rust bindings
 */

import * as flatbuffers from 'flatbuffers';
import { PhoneticAtlas } from './generated/phonetic-atlas';
import type { AtlasMetadata } from './generated/atlas-metadata';
import type { PhonemeEntry } from './generated/phoneme-entry';
import type { FeatureVector } from './generated/feature-vector';

/**
 * Error thrown when the atlas data is structurally corrupted.
 * This is a distinct error type so consumers can differentiate
 * "phoneme not found" from "atlas data corrupt".
 */
export class AtlasCorruptionError extends Error {
  constructor(message: string) {
    super(`[PhoneticAtlas] ${message}`);
    this.name = 'AtlasCorruptionError';
  }
}

// ---------------------------------------------------------------------------
// Public types — re-exported for consumer convenience
// ---------------------------------------------------------------------------
export type { AtlasMetadata, PhonemeEntry, FeatureVector };

/**
 * Feature value enum matching the FlatBuffers schema.
 *
 *   0 = Unspecified / not applicable  (panphon "0")
 *   1 = Positively specified          (panphon "+")
 *   2 = Negatively specified          (panphon "-")
 */
export enum FeatureValue {
  Unspecified = 0,
  Positive = 1,
  Negative = 2,
}

/**
 * Decoded phoneme entry with zero-copy feature vector access.
 */
export interface PhonemeInfo {
  /** IPA symbol (e.g., "p", "aɪ"), NFC-normalized */
  readonly ipa: string;
  /** True if this is a base phoneme */
  readonly isBase: boolean;
  /** Phonological feature vector (zero-copy struct access) */
  readonly features: FeatureVector;
}

/**
 * Runtime metadata extracted from the FlatBuffers atlas header.
 */
export interface AtlasInfo {
  readonly sourceName: string;
  readonly sourceVersion: string;
  readonly generatorName: string;
  readonly generatorVersion: string;
  readonly generatedAt: string;
  readonly contentHash: string;
  readonly totalSegments: number;
  readonly totalBases: number;
  readonly featureCount: number;
}

/**
 * Phonetic Atlas — zero-copy access to the phoneme feature vector set.
 *
 * Loaded once at process/worker init. The internal `_atlas` reference
 * holds the FlatBuffers root table; all access methods read directly
 * from the underlying ArrayBuffer without intermediate allocations.
 *
 * IPA LOOKUP NORMALIZATION:
 *   All IPA strings are stored and queried in NFC form. This means:
 *     atlas.get("ä") === atlas.get("a\u0308")  // both find the same entry
 *   The index stores NFC keys; the lookup function normalizes input to NFC.
 */
export class PhoneticAtlasIndex {
  private readonly _atlas: PhoneticAtlas;
  /** IPA-symbol (NFC) → index lookup. Built once at load time. */
  private readonly _index: Map<string, number>;

  private constructor(atlas: PhoneticAtlas) {
    this._atlas = atlas;

    // Build IPA → index map for O(1) lookup. The phoneme table is small
    // (<7000 entries), so Map memory overhead (~48 bytes per entry) is
    // negligible compared to the speed gain on every stress lookup.
    const n = this._atlas.phonemesLength();
    const idx = new Map<string, number>();
    for (let i = 0; i < n; i++) {
      const p = this._atlas.phonemes(i);
      if (!p) continue;

      const rawIpa = p.ipa();
      if (rawIpa === null || rawIpa === undefined) {
        // C5: Null IPA in the binary is data corruption. The schema
        // marks ipa as (required), so this should never happen with a
        // valid atlas. We throw rather than silently skipping — a
        // corrupted atlas should fail loudly, not silently lose entries.
        throw new AtlasCorruptionError(
          `Null IPA at phoneme index ${i}. Atlas binary is corrupt.`,
        );
      }

      // C1: NFC normalization — all IPA symbols stored as NFC.
      // Panphon source data uses NFC for precomposed diacritics.
      // Without normalization, lookups for composed vs decomposed
      // forms of the same symbol produce false negatives.
      const nfcIpa = rawIpa.normalize('NFC');
      if (idx.has(nfcIpa)) {
        // Duplicate IPA after NFC normalization. This can happen if
        // the source data has two different entries that normalize
        // to the same string. Last-write wins — log a warning.
        console.warn(
          `[PhoneticAtlas] Duplicate NFC-normalized IPA: "${nfcIpa}" ` +
          `at index ${i} (overwrites index ${idx.get(nfcIpa)})`,
        );
      }
      idx.set(nfcIpa, i);
    }
    this._index = idx;
  }

  // -----------------------------------------------------------------------
  // Factory
  // -----------------------------------------------------------------------

  /**
   * Create a PhoneticAtlasIndex from an ArrayBuffer containing the
   * compiled FlatBuffers binary.
   *
   * @param buffer - The raw bytes of phonetic_atlas.bin
   * @returns A ready-to-use index
   * @throws AtlasCorruptionError if the buffer lacks the "PHAT" identifier
   */
  static fromBuffer(buffer: ArrayBuffer | Uint8Array): PhoneticAtlasIndex {
    const bytes = buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);

    // C7: Validate the "PHAT" file identifier BEFORE parsing.
    // Without this check, garbage bytes would be parsed as an atlas,
    // producing silently wrong results from random offsets.
    if (!bytes || bytes.length < 8) {
      throw new AtlasCorruptionError(
        'Buffer too small for FlatBuffers file identifier. Expected at least 8 bytes.',
      );
    }
    if (!PhoneticAtlas.bufferHasIdentifier(new flatbuffers.ByteBuffer(bytes))) {
      throw new AtlasCorruptionError(
        'Buffer does not have the "PHAT" file identifier. ' +
        'Expected a valid Phonetic Atlas FlatBuffers binary.',
      );
    }

    const bb = new flatbuffers.ByteBuffer(bytes);
    const atlas = PhoneticAtlas.getRootAsPhoneticAtlas(bb);
    return new PhoneticAtlasIndex(atlas);
  }

  /**
   * Verify the buffer has the correct "PHAT" file identifier.
   *
   * This is a cheap check (just 4 bytes at a known offset).
   */
  static hasValidIdentifier(buffer: ArrayBuffer | Uint8Array): boolean {
    const bytes = buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);
    if (bytes.length < 8) return false;
    return PhoneticAtlas.bufferHasIdentifier(new flatbuffers.ByteBuffer(bytes));
  }

  // -----------------------------------------------------------------------
  // Metadata
  // -----------------------------------------------------------------------

  /** Runtime metadata from the atlas header. */
  get metadata(): AtlasInfo | null {
    const m = this._atlas.metadata();
    if (!m) return null;
    return {
      sourceName: m.sourceName() ?? '',
      sourceVersion: m.sourceVersion() ?? '',
      generatorName: m.generatorName() ?? '',
      generatorVersion: m.generatorVersion() ?? '',
      generatedAt: m.generatedAt() ?? '',
      contentHash: m.contentHash() ?? '',
      totalSegments: m.totalSegments(),
      totalBases: m.totalBases(),
      featureCount: m.featuresLength(),
    };
  }

  // -----------------------------------------------------------------------
  // Lookup
  // -----------------------------------------------------------------------

  /**
   * Look up a phoneme by its IPA symbol.
   *
   * The input IPA is normalized to NFC before lookup, so composed and
   * decomposed forms of the same symbol are treated as equivalent.
   *
   * @returns PhonemeInfo with zero-copy feature vector, or null if not found
   */
  get(ipa: string): PhonemeInfo | null {
    // C1: NFC-normalize the query string to match the index keys
    const normalized = ipa.normalize('NFC');
    const i = this._index.get(normalized);
    if (i === undefined) return null;
    return this._entryAt(i);
  }

  /**
   * Get a phoneme by its positional index in the atlas.
   *
   * Useful for iterating all phonemes or for fast random access
   * when the index is already known.
   */
  at(index: number): PhonemeInfo | null {
    if (index < 0 || index >= this._atlas.phonemesLength()) return null;
    return this._entryAt(index);
  }

  /**
   * Iterate all phonemes in the atlas.
   *
   * Yields PhonemeInfo objects with zero-copy feature vector access.
   * The returned objects are valid only during iteration; they share
   * the underlying ArrayBuffer and do not allocate copies.
   *
   * @yields PhonemeInfo for each valid entry.
   * @throws AtlasCorruptionError if any entry has null features.
   */
  *entries(): IterableIterator<PhonemeInfo> {
    const n = this._atlas.phonemesLength();
    for (let i = 0; i < n; i++) {
      // C6: Don't skip corrupted entries — fail loud so the consumer
      // knows the atlas is unreliable rather than silently losing data.
      yield this._entryAtOrThrow(i);
    }
  }

  /** Total number of phoneme entries in the atlas. */
  get size(): number {
    return this._atlas.phonemesLength();
  }

  // -----------------------------------------------------------------------
  // Internal
  // -----------------------------------------------------------------------

  /**
   * Safe accessor — returns null for out-of-bounds, but throws on
   * structural corruption (null features).
   *
   * This is the "lenient" version used by get() and at() — null means
   * "phoneme doesn't exist", which is a normal case for user queries.
   * But if the phoneme DOES exist and its data is corrupt, we throw.
   */
  private _entryAt(index: number): PhonemeInfo | null {
    const p = this._atlas.phonemes(index);
    if (!p) return null;

    // C6: Null features = corrupted entry. We throw rather than returning
    // null because null would be indistinguishable from "phoneme not found"
    // to the caller. A corrupted atlas is a hard failure, not a soft miss.
    const fv = p.features();
    if (!fv) {
      throw new AtlasCorruptionError(
        `Null feature vector at phoneme index ${index}. Atlas binary is corrupt.`,
      );
    }

    // C1: NFC-normalize the returned IPA so that all surfaces are NFC
    const rawIpa = p.ipa();
    return {
      ipa: rawIpa ? rawIpa.normalize('NFC') : '',
      isBase: p.isBase(),
      features: fv,
    };
  }

  /**
   * Strict accessor — used by entries() iteration. Throws on ANY
   * null (corrupt entry or corrupt feature vector).
   */
  private _entryAtOrThrow(index: number): PhonemeInfo {
    const p = this._atlas.phonemes(index);
    if (!p) {
      throw new AtlasCorruptionError(
        `Null phoneme entry at index ${index}. Atlas binary is corrupt.`,
      );
    }
    const fv = p.features();
    if (!fv) {
      throw new AtlasCorruptionError(
        `Null feature vector at phoneme index ${index}. Atlas binary is corrupt.`,
      );
    }
    const rawIpa = p.ipa();
    return {
      ipa: rawIpa ? rawIpa.normalize('NFC') : '',
      isBase: p.isBase(),
      features: fv,
    };
  }
}

File: packages/phonetics-core/src/atlas/load_atlas.py
"""
Phonetic Atlas Loader — Python entry point for analytics-pipeline
==================================================================

Loads the compiled FlatBuffers Phonetic Atlas (.bin) via memory-mapped
file (mmap) and provides typed access to phoneme feature vectors with
zero-copy reads.

WHY mmap:
  - The atlas is a static, read-heavy asset (~305 KB, 6367 entries).
  - mmap maps the file into the process virtual address space — the OS
    pages in data on demand, and no Python allocation occurs for the
    raw buffer.
  - FeatureVector is a struct (inline, fixed 24 bytes) — reading a
    single field reads directly from mapped memory.

WHY a separate module (#2 in the guide):
  - flatc-generated Python bindings provide GetRootAs / low-level
    access. This module wraps them in a convenient PhoneticAtlasIndex
    with dict-based IPA→entry lookup.
  - The analytics pipeline (analytics-pipeline/src/core/numpy_calc.py)
    imports this to transform feature vectors into C-contiguous NumPy
    arrays.

USAGE:
    from atlas.load_atlas import load_phonetic_atlas
    atlas = load_phonetic_atlas("path/to/phonetic_atlas.bin")
    phoneme = atlas.get("p")
    if phoneme:
        print(phoneme.features.Cons())  # 1 = [+consonantal]

See architecture.md §5 "Phonetic Atlas — the cross-language data asset"
"""

from __future__ import annotations

import mmap
import os
import unicodedata
from typing import Optional


class AtlasCorruptionError(Exception):
    """Raised when the atlas binary is structurally corrupted."""
    pass

# ---------------------------------------------------------------------------
# FlatBuffers Python bindings (flatc-generated from phonetic_atlas.fbs)
#
# These are imported from the dist/python/ directory of the phonetics-core
# package. The import path works when:
#   - Running from within the monorepo (workspace path resolves it)
#   - The package is installed via pip -e or the dist-info is present
# ---------------------------------------------------------------------------
try:
    from PhoneticAtlas import PhoneticAtlas  # type: ignore[import-not-found]
    from PhonemeEntry import PhonemeEntry  # type: ignore[import-not-found]
    from FeatureVector import FeatureVector  # type: ignore[import-not-found]
    from AtlasMetadata import AtlasMetadata  # type: ignore[import-not-found]
except ImportError:
    # Fallback: try to find generated bindings relative to this file
    _PKG_DIR = os.path.normpath(
        os.path.join(os.path.dirname(__file__), "..", "..", "dist", "python")
    )
    if os.path.isdir(_PKG_DIR):
        import sys
        sys.path.insert(0, _PKG_DIR)
        from PhoneticAtlas import PhoneticAtlas  # type: ignore[import-not-found]
        from PhonemeEntry import PhonemeEntry  # type: ignore[import-not-found]
        from FeatureVector import FeatureVector  # type: ignore[import-not-found]
        from AtlasMetadata import AtlasMetadata  # type: ignore[import-not-found]
    else:
        raise ImportError(
            "FlatBuffers generated Python bindings not found. "
            "Run: pnpm --filter @poetry/phonetics-core codegen"
        )


class PhonemeInfo:
    """Decoded phoneme entry with zero-copy feature vector access."""

    __slots__ = ("ipa", "is_base", "features")

    def __init__(
        self,
        ipa: str,
        is_base: bool,
        features: FeatureVector,
    ) -> None:
        self.ipa = ipa
        self.is_base = is_base
        self.features = features  # Zero-copy struct — reads from mmap'd memory


class AtlasMetadataInfo:
    """Metadata extracted from the FlatBuffers atlas header."""

    __slots__ = (
        "source_name", "source_version",
        "generator_name", "generator_version",
        "generated_at", "content_hash",
        "total_segments", "total_bases",
        "feature_count",
    )

    def __init__(self, meta: AtlasMetadata) -> None:
        def _s(b: Optional[bytes]) -> str:
            return b.decode("utf-8") if b else ""
        self.source_name = _s(meta.SourceName())
        self.source_version = _s(meta.SourceVersion())
        self.generator_name = _s(meta.GeneratorName())
        self.generator_version = _s(meta.GeneratorVersion())
        self.generated_at = _s(meta.GeneratedAt())
        self.content_hash = _s(meta.ContentHash())
        self.total_segments = meta.TotalSegments()
        self.total_bases = meta.TotalBases()
        self.feature_count = meta.FeaturesLength()


class PhoneticAtlasIndex:
    """
    Phonetic Atlas — zero-copy access to the phoneme feature vector set.

    Loaded once at process start. The internal `_atlas` reference holds
    the FlatBuffers root table; all access methods read directly from
    the memory-mapped buffer without intermediate allocations.
    """

    def __init__(self, atlas: PhoneticAtlas) -> None:
        self._atlas = atlas
        # Build IPA → index dict for O(1) lookup
        # C1: All IPA symbols are NFC-normalized. This ensures composed
        # and decomposed forms of the same symbol map to the same entry.
        n = self._atlas.PhonemesLength()
        self._index: dict[str, int] = {}
        for i in range(n):
            p = self._atlas.Phonemes(i)
            if not p:
                continue
            ipa_bytes = p.Ipa()
            if ipa_bytes is None:
                # C5: Null IPA = data corruption. Schema marks ipa as
                # (required), so this should never happen in valid data.
                raise AtlasCorruptionError(
                    f"Null IPA at phoneme index {i}. Atlas binary is corrupt."
                )
            ipa = unicodedata.normalize("NFC", ipa_bytes.decode("utf-8"))
            if ipa in self._index:
                import warnings
                warnings.warn(
                    f"Duplicate NFC-normalized IPA: '{ipa}' at index {i} "
                    f"(overwrites index {self._index[ipa]})"
                )
            self._index[ipa] = i

    # -------------------------------------------------------------------
    # Factory
    # -------------------------------------------------------------------

    @classmethod
    def from_path(cls, path: str) -> "PhoneticAtlasIndex":
        """
        Load the atlas from a binary file using memory-mapped I/O.

        The file is mapped into the process address space (mmap) — the
        OS handles paging, and no Python buffer allocation occurs for
        the raw data. The mmap stays valid for the lifetime of this
        object.

        Raises:
            AtlasCorruptionError: if the file lacks the "PHAT" identifier
                or is too small to be a valid FlatBuffer.
        """
        with open(path, "rb") as f:
            mm = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ)

        # C7: Validate "PHAT" file identifier before parsing
        if len(mm) < 8:
            mm.close()
            raise AtlasCorruptionError(
                f"File too small for FlatBuffers: {path} ({len(mm)} bytes)."
            )
        if not PhoneticAtlas.PhoneticAtlasBufferHasIdentifier(mm, 0):
            mm.close()
            raise AtlasCorruptionError(
                f"File does not have the 'PHAT' identifier: {path}"
            )

        atlas = PhoneticAtlas.GetRootAsPhoneticAtlas(mm, 0)
        return cls(atlas)

    # -------------------------------------------------------------------
    # Metadata
    # -------------------------------------------------------------------

    @property
    def metadata(self) -> Optional[AtlasMetadataInfo]:
        m = self._atlas.Metadata()
        return AtlasMetadataInfo(m) if m else None

    # -------------------------------------------------------------------
    # Lookup
    # -------------------------------------------------------------------

    def get(self, ipa: str) -> Optional[PhonemeInfo]:
        """Look up a phoneme by its IPA symbol. Returns None if not found.

        The input IPA is NFC-normalized, so composed and decomposed forms
        of the same symbol (e.g., "ä" vs "a\\u0308") are equivalent.
        """
        # C1: NFC-normalize the query to match index keys
        normalized = unicodedata.normalize("NFC", ipa)
        i = self._index.get(normalized)
        if i is None:
            return None
        return self._entry_at(i)

    def at(self, index: int) -> Optional[PhonemeInfo]:
        """Get a phoneme by its positional index in the atlas."""
        if index < 0 or index >= self._atlas.PhonemesLength():
            return None
        return self._entry_at(index)

    def __len__(self) -> int:
        return self._atlas.PhonemesLength()

    def __iter__(self):
        n = self._atlas.PhonemesLength()
        for i in range(n):
            # C6: Don't skip corrupted entries — fail loud.
            # The strict accessor throws on null phoneme or null features.
            yield self._entry_at_strict(i)

    # -------------------------------------------------------------------
    # Internal
    # -------------------------------------------------------------------

    def _entry_at(self, index: int) -> Optional[PhonemeInfo]:
        """Lenient accessor — returns None for out-of-bounds, but throws
        on structural corruption (null features, null IPA).
        This is used by get() and at() — null means "phoneme doesn't
        exist", which is normal. But if data is corrupt, we throw."""
        p = self._atlas.Phonemes(index)
        if p is None:
            return None
        # C6: Null features = corruption. Throw distinctly so the caller
        # can differentiate "phoneme not found" from "atlas corrupt".
        fv = p.Features()
        if fv is None:
            raise AtlasCorruptionError(
                f"Null feature vector at phoneme index {index}. "
                f"Atlas binary is corrupt."
            )
        ipa_bytes = p.Ipa()
        if ipa_bytes is None:
            raise AtlasCorruptionError(
                f"Null IPA at phoneme index {index}. Atlas binary is corrupt."
            )
        # C1: NFC-normalize the IPA before returning
        ipa = unicodedata.normalize("NFC", ipa_bytes.decode("utf-8"))
        return PhonemeInfo(
            ipa=ipa,
            is_base=p.IsBase(),
            features=fv,
        )

    def _entry_at_strict(self, index: int) -> PhonemeInfo:
        """Strict accessor — used by iteration. Throws on ANY null:
        corrupt entry or corrupt feature vector."""
        p = self._atlas.Phonemes(index)
        if p is None:
            raise AtlasCorruptionError(
                f"Null phoneme entry at index {index}. Atlas binary is corrupt."
            )
        fv = p.Features()
        if fv is None:
            raise AtlasCorruptionError(
                f"Null feature vector at phoneme index {index}. "
                f"Atlas binary is corrupt."
            )
        ipa_bytes = p.Ipa()
        if ipa_bytes is None:
            raise AtlasCorruptionError(
                f"Null IPA at phoneme index {index}. Atlas binary is corrupt."
            )
        ipa = unicodedata.normalize("NFC", ipa_bytes.decode("utf-8"))
        return PhonemeInfo(
            ipa=ipa,
            is_base=p.IsBase(),
            features=fv,
        )


# ---------------------------------------------------------------------------
# Convenience factory
# ---------------------------------------------------------------------------

def load_phonetic_atlas(path: Optional[str] = None) -> PhoneticAtlasIndex:
    """
    Load the Phonetic Atlas from disk.

    Args:
        path: Path to phonetic_atlas.bin. If None, auto-resolves relative
              to this package's dist/ directory.

    Returns:
        A ready-to-use PhoneticAtlasIndex.
    """
    if path is None:
        path = os.path.normpath(
            os.path.join(
                os.path.dirname(__file__),
                "..", "..", "dist", "phonetic_atlas.bin"
            )
        )
    return PhoneticAtlasIndex.from_path(path)
